docs/api-levels.rst

```diff
@@ -0,0 +1,87 @@
+
+.. danger::
+
+   At this point, python-gitlab will no longer take care of URL-encoding and other transformations
+   needed to correctly pass API parameter types. You have to construct these correctly yourself.
+
+.. important::
+
+   If you've found yourself at this section because of an endpoint not yet implemented in
+   the library - please consider opening a pull request implementing the resource or at
+   least filing an issue so we can track progress.
+
+   High-quality pull requests for standard endpoints that pass CI and include unit tests and
+   documentation are easy to review, and can land quickly with monthly releases. If you ask,
+   we can also trigger a new release, so you and everyone benefits from the contribution right away!
+
+Managers and objects call specific HTTP methods to fetch or send data to the server. These methods
+can be invoked directly to access endpoints not currently implemented by the client. This essentially
+gives you some level of usability for any endpoint the moment it is available on your GitLab instance.
+
+These methods can be accessed directly via the ``Gitlab`` instance (e.g. ``gl.http_get()``), or via an
+object's manager (e.g. ``project.manager.gitlab.http_get()``), if the ``Gitlab`` instance is not available
+in the current context.
+
+For example, if you'd like to access GitLab's `undocumented latest pipeline endpoint
+<https://gitlab.com/gitlab-org/gitlab/-/blob/5e2a61166d2a033d3fd1eb4c09d896ed19a57e60/lib/api/ci/pipelines.rb#L97>`__,
+you can do so by calling ``http_get()`` with the path to the endpoint:
+
+.. code-block:: python
+
+    >>> gl = gitlab.Gitlab(private_token=private_token)
+    >>>
+    >>> pipeline = gl.http_get("/projects/gitlab-org%2Fgitlab/pipelines/latest")
+    >>> pipeline["id"]
+    449070256
+
+The available methods are:
+
+* ``http_get()``
+* ``http_post()``
+* ``http_put()``
+* ``http_delete()``
+* ``http_list()`` (a wrapper around ``http_get`` handling pagination, including with lazy generators)
+* ``http_head()`` (only returns the header dictionary)
+
+Lower-lower-level API - HTTP requests
+=====================================
+
+.. important::
+
+    This is mostly intended for internal use in python-gitlab and may have a less stable interface than
+    higher-level APIs.
+
+At the lowest level, HTTP methods call ``http_request()``, which performs the actual request and takes
+care of details such as timeouts, retries, and handling rate-limits.
+
+This method can be invoked directly to or customize this behavior for a single request, or to call custom
+HTTP methods not currently implemented in the library - while still making use of all of the client's
+options and authentication methods.
+
+For example, if for whatever reason you want to fetch allowed methods for an endpoint at runtime:
+
+.. code-block:: python
+
+    >>> gl = gitlab.Gitlab(private_token=private_token)
+    >>>
+    >>> response = gl.http_request("OPTIONS", "/projects")
```

Comment:
Can keyword arguments be used here? As I'm not 100% sure what the arguments are at first glance.

Reply:
I've looked at all the well-known http libraries and they all have the `request()` interface done like this (stdlib's http.client, urllib3, requests, httpx, httpcore, aiohttp), with those that enforce keyword-only args doing it after the method+url pair. So I thought this would make sense for people looking at lower-level APIs. I've added another example where we add some kwargs. If you feel really strongly about this I can still change it? :)

Reply:
I ended up adding this in https://github.com/python-gitlab/python-gitlab/pull/2230/commits/af3d98bd2d13e77fe26fd58f626759e8515a3f3e.